### PR TITLE
[microcosm] Async observables

### PR DESCRIPTION
**Pull request by nhunzaker**
*Review notes:* It's probably best to view this in split-view

### What is it

This commit updates the Observable abstraction in Microcosm such that it is asynchronous.

Updates to observables are queued with a universal scheduler. Until this scheduler "flushes" an observable, updates are buffered within an internal list of updates.

### Why 

The scheduler dispatches updates as fast as the browser will allow it to, using requestIdleCallback. When requestIdleCallback is not provided, it uses setTimeout.

This means that Microcosm's scheduler does not block the browser. Work might be slow, but the browser never suffers.

### CODE REVIEW

File: packages/microcosm/src/scheduler.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+    this.scheduled = true
+    this.method(this.tick)
+  }
+
+  tick(idleDeadline) {
+    try {
+      while (this.queue.length && idleDeadline.timeRemaining() > 0) {
+        this.queue.shift()()
+      }
```
Comment (nhunzaker) [inline packages/microcosm/src/scheduler.js]: Using [requestIdleCallback](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestIdleCallback)'s API, this says:

> Run as many queued jobs as possible until the browser says it has to do something, like paint.

File: packages/microcosm/src/scheduler.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    }
+  }
+
+  flush() {
+    this.tick(untilEmpty)
+  }
+
+  onError(fn) {
+    this.errorCallbacks.push(fn)
+  }
```
Comment (nhunzaker) [inline packages/microcosm/src/scheduler.js]: This lets you subscribe, globally, to uncaught errors executing jobs. Jobs is vague here. This is basically any state update within a Microcosm, Domain, Effect, or Action. Period.

Pretty neat :)

File: packages/microcosm/test/setup.js [hunk 3]
```diff
@@ -1,6 +1,10 @@
-import { get } from 'microcosm'
+import { get, scheduler } from 'microcosm'
 import expect from 'expect'
 
+scheduler().onError(error => {
+  throw error instanceof Error ? error : new Error(error)
+})
```
Comment (nhunzaker) [inline packages/microcosm/test/setup.js]: Says: while testing, if a job has an uncaught exception, fail immediately.

File: packages/microcosm/test/unit/observable.test.js [hunk 4]
```diff
@@ -14,10 +14,28 @@ describe('Observable', function() {
...
+        throw 'Terrible things'
+      })
+
+      observable.subscribe({
+        error(reason) {
+          expect(reason).toEqual('Terrible things')
+        }
+      })
+
+      scheduler().flush()
```
Comment (nhunzaker) [inline packages/microcosm/test/unit/observable.test.js]: Flush is a nice utility for forcing all updates to send out. You could also use async/await like `await scheduler`, but this has the nice benefit of test execution time.

File: packages/microcosm/src/observable.js [hunk 5]
```diff
@@ -181,105 +159,136 @@ export class Observable {
...
+function flushSubscription(subscription: Subscription): void {
+  let queue = subscription._queue
 
-  observer.complete()
+  subscription._queue = []
+  subscription._state = WORKING
 
-  cleanupSubscription(subscription)
-}
+  for (let i = 0; i < queue.length; ++i) {
```
Comment (zporter) [inline packages/microcosm/src/observable.js]: would `++i` increment `i` to `1` before the block is called?

File: packages/microcosm/src/scheduler.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    )
+
+    this.queue.push(cb)
+    this.schedule()
+  }
+
+  schedule() {
+    if (this.scheduled) {
+      return
+    }
```
Comment (zporter) [inline packages/microcosm/src/scheduler.js]: nice catch 👏

File: packages/microcosm/src/observable.js (hunk 5 above)
Comment (nhunzaker) [reply packages/microcosm/src/observable.js]: Nope! The third statement is the post-condition of the loop. This is just confusing. I'll replace it with `i++`, which I should have had to begin with.

File: packages/microcosm/test/unit/observable.test.js [hunk 7]
```diff
@@ -14,10 +14,28 @@ describe('Observable', function() {
           expect(reason).toEqual('Terrible things')
         }
       })
+
+      scheduler().flush()
+    })
+
+    it('cleans ups any enqueued tasks on unsubscribe', () => {
```
Comment (zporter) [inline packages/microcosm/test/unit/observable.test.js]: nitpick: `cleans up`

Comment (nhunzaker) [reply packages/microcosm/test/unit/observable.test.js]: I keep all my ups clena.

**Comment by nhunzaker** [comment]
Thanks, @zporter !

**Pull request merged**